### PR TITLE
Feat: settime chat command

### PR DESCRIPTION
**Pull request by ToeKneeRED**
It is about the same if not the same as the [previous PR](https://github.com/tiltedphoques/TiltedEvolution/pull/510) for it from cosi.
Adds a `/settime` chat command that can be used on any server that is not public. 

Ideally, there would be some feedback from the server that it succeeded or failed besides the time change (e.g. a localized chat message, but that turned out to be too complex for me).

EDIT:
Since admins should be the only ones allowed to use the /settime chat command, there are now two different ways of becoming an admin; either by joining the server using the admin password specified in the STServer.ini file, or using the new /AddAdmin console command.
Unfortunately, spaces in names seem to be nearly impossible to deal with without replacing the characters and re-checking for the name over and over again. For the sake of this PR, you can't /AddAdmin or /RemoveAdmin on a name with an underscore in it as that is what is used to replace since spaces are more likely to be in names rather than underscores.

Three new console commands:

/AddAdmin <username>: add the player with the username given to the adminSessions Set
ex.
/AddAdmin Toe_Knee = Toe Knee
/AddAdmin ToeKnee = ToeKnee
/RemoveAdmin <username>: remove the player with the username given from the adminSessions Set
/admins: list out all admins

The PR also now includes a fix for a crash with the ConsoleRegistry's manual resize once it reaches a size() of 10 (it tried to erase m_commandHistory.end()).

### CODE REVIEW

**Comment by ToeKneeRED** [comment]
Since admins should be the only ones allowed to use the `/settime` chat command, there are now two different ways of becoming an admin; either by joining the server using the admin password specified in the STServer.ini file, or using the new `/AddAdmin` console command. 
Unfortunately, spaces in names seem to be nearly impossible to deal with without replacing the characters and re-checking for the name over and over again. For the sake of this PR, you can't `/AddAdmin` or `/RemoveAdmin` on a name with an underscore in it as that is what is used to replace since spaces are more likely to be in names rather than underscores.

Three new console commands: 
- `/AddAdmin <username>`: add the player with the username given to the adminSessions Set
 ex. 
`/AddAdmin Toe_Knee` = Toe Knee 
`/AddAdmin ToeKnee` = ToeKnee
- `/RemoveAdmin <username>`: remove the player with the username given from the adminSessions Set
- `/admins`: list out all admins
     
The PR also now includes a fix for a crash with the ConsoleRegistry's manual resize once it reaches a size() of 10 (it tried to erase m_commandHistory.end()).

File: Code/server/GameServer.cpp [hunk 1]
```diff
@@ -383,6 +391,129 @@ void GameServer::BindServerCommands()
                 out->error("Day must be between 0 and 31, month must be between 0 and 11, and year must be between 0 and 999.");
             }
         });
+
+    m_commands.RegisterCommand<std::string>(
+        "AddAdmin", "Add admin privileges from player", [&](Console::ArgStack& aStack) {
```
Comment (Force67) [inline Code/server/GameServer.cpp:396]: to player

File: Code/server/GameServer.cpp [hunk 2]
```diff
@@ -383,6 +391,129 @@ void GameServer::BindServerCommands()
...
+                    out->info("{} admin privileges added", cUsername.c_str());
+                    playerFound = true;
+                    return;
+                }
+            });
+
+            if (!playerFound)
+            {
+                // space in username handling
+                String backupUsername = cUsername;
```
Comment (Force67) [inline Code/server/GameServer.cpp:416]: Should factor out the user name sanitizing / cleaning into a separate utility fn.

File: Code/server/GameServer.cpp [hunk 3]
```diff
@@ -383,6 +391,129 @@ void GameServer::BindServerCommands()
...
+    m_commands.RegisterCommand<std::string>(
+        "RemoveAdmin", "Remove admin privileges from player", [&](Console::ArgStack& aStack) {
+            auto out = spdlog::get("ConOut");
+
+            const auto& cUsername = aStack.Pop<String>();
+            bool playerFound = false;
+
+            for (const auto& cAdmin : m_adminSessions)
+            {
+                Player* pPlayer = PlayerManager::Get()->GetByConnectionId(cAdmin);
```
Comment (Force67) [inline Code/server/GameServer.cpp:449]: Probably the get user by name stuff too. maybe have a fn there to get a user by name.

File: Code/server/GameServer.cpp [hunk 4]
```diff
@@ -383,6 +391,129 @@ void GameServer::BindServerCommands()
...
+                return;
+            }
+
+            String output = "Admins: ";
+            bool _first = true;
+
+            for (const auto& cAdmin : m_adminSessions)
+            {
+                const auto& cUsername = PlayerManager::Get()->GetByConnectionId(cAdmin)->GetUsername();
+
```
Comment (Force67) [inline Code/server/GameServer.cpp:503]: Even though unlikely, we should still handle if GetConnectionId somehow returns null

File: Code/client/Services/TransportService.h [hunk 5]
```diff
@@ -32,10 +34,12 @@ struct TransportService : Client
 
     [[nodiscard]] bool IsOnline() const noexcept { return m_connected; }
     void SetServerPassword(const std::string& acPassword) noexcept { m_serverPassword = acPassword; }
+    const uint32_t& GetLocalPlayerId() const noexcept { return m_localPlayerId; }
 
 protected:
     // Event handlers
     void HandleUpdate(const UpdateEvent& acEvent) noexcept;
+    void HandleConnect(const ConnectedEvent& acEvent) noexcept;
```
Comment (Force67) [inline Code/client/Services/TransportService.h:42]: We should for safety reasons tombstone the localplayerid on disconnect. e.g. set it to invalid.

File: Code/server/GameServer.cpp [hunk 6]
```diff
@@ -747,8 +879,14 @@ void GameServer::HandleAuthenticationRequest(const ConnectionId_t aConnectionId,
     }
 
     // check if the proper server password was supplied.
-    if (acRequest->Token == sPassword.value())
+    if (acRequest->Token == sPassword.value() || (acRequest->Token == sAdminPassword.value() && !sAdminPassword.empty()))
```
Comment (Force67) [inline Code/server/GameServer.cpp:882]: might be nicer to assign the result of the second expr to a boolean then we can also use the result below

File: Code/server/GameServer.h [hunk 7]
```diff
@@ -78,6 +78,14 @@ struct GameServer final : Server
     {
         return m_isPasswordProtected;
     }
+    void SetPublic(bool aIsPublic) noexcept
+    {
+        m_isPublic = aIsPublic;
+    }
+    bool IsPublic() const noexcept
```
Comment (Force67) [inline Code/server/GameServer.h:85]: nodiscard

File: Code/server/GameServer.h [hunk 8]
```diff
@@ -78,6 +78,14 @@ struct GameServer final : Server
     {
         return m_isPasswordProtected;
     }
+    void SetPublic(bool aIsPublic) noexcept
```
Comment (Force67) [inline Code/server/GameServer.h:81]: There is no need for this. the value if a server is "public" is contained by bAnnounceServer convar. if you need it from another cpp file, you can extern Convar<bool> name it.

**Review by Force67 — CHANGES_REQUESTED** [review]
Also please run clang format python script (tools/scripts/clang_format.py) over your changes.

**Comment by ToeKneeRED** [comment]
clang-format'd all files except those in the scripting expansion because I accidentally rebased instead of merge and it wants to re-add those files

File: Code/server/GameServer.cpp (hunk 1 above)
Comment (ToeKneeRED) [reply Code/server/GameServer.cpp]: Fixed

File: Code/client/Services/TransportService.h (hunk 5 above)
Comment (ToeKneeRED) [reply Code/client/Services/TransportService.h]: Fixed with new HandleDisconnected

File: Code/server/GameServer.cpp (hunk 3 above)
Comment (ToeKneeRED) [reply Code/server/GameServer.cpp]: Added two things for this, GameServer::GetAdminByUsername and PlayerManager::GetByUsername
Reworked that part to adapt it in as well

File: Code/server/GameServer.cpp (hunk 4 above)
Comment (ToeKneeRED) [reply Code/server/GameServer.cpp:494]: Now has a nullptr check and prints to console if m_adminSessions contains a session it shouldn't

File: Code/server/GameServer.cpp (hunk 6 above)
Comment (ToeKneeRED) [reply Code/server/GameServer.cpp]: Fixed with additional bool above

File: Code/server/GameServer.h (hunk 8 above)
Comment (ToeKneeRED) [reply Code/server/GameServer.h]: Removed it, and the reference to it that was in ServerListService

**Review by Force67 — APPROVED** [review]

**Comment by rfortier** [comment]
I'm late to the party since this was just merged. But I added it to the latest Modded Animation pre-release binary build, after verifying (shocker, it just works). So we can get a few more testers hands-on before a Master release.

**Pull request merged**